### PR TITLE
 fix(*): gRPC and cJSON's behavior to empty array

### LINKED ISSUE #10801: [gRPC-gateway] Empty arrays are decoded as `{}` (which should be `[]`)
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Kong version (`$ kong version`)

\*

### Current Behavior

When a gRPC service returns an empty array, it's encoded as `{}`  in the JSON body.

### Expected Behavior

It should be encoded as `[]`.

### Steps To Reproduce

```protobuf
//... some service
  rpc Echo(EchoMsg) returns (EchoMsg) {
    option (google.api.http) = {
      post: "/v1/echo"
      body: "*"
    };
  }

//...
message EchoMsg {
  repeated string array = 1;
  string nullable = 2;
}
```

```go
func (s *server) Echo(ctx context.Context, in *pb.EchoMsg) (*pb.EchoMsg, error) {
	return in, nil
}
```

```lua
local req_body = { array = {}, nullable = "ahaha" }
local res, _ = proxy_client:post("/v1/echo", {
   headers = { ["Content-Type"] = "application/json" },
    body = req_body,
})
```

`res:read_body()` would return something like `{ "array": {}, "nullable": "ahaha"}`.

### Anything else?

This issue is created for visibility for an internal ticket(FTI-5002).

### PR DESCRIPTION
**Pull request by StarlightIbuki**
### Summary

Lua does not distinguish empty arrays and tables, which makes empty array decoded from gRPC gateway be encoded as `{}` by cJSON.
The fix makes use of cjson.empty_array_mt to hint the cJSON.
Special thanks to @starwing, who spent their public holidays adapting new features for this issue.

### Checklist

- [x] The Pull Request has tests
- [x] There's an entry in the CHANGELOG
- [ ] N/A <s>There is a user-facing docs PR against https://github.com/Kong/docs.konghq.com - PUT DOCS PR HERE</s>

### Full changelog

* Bumped lua-protobuf
* Use `cjson.empty_array_t` for arrays
* Tests

### Issue reference

https://github.com/starwing/lua-protobuf/pull/240
Fix FTI-5002
Fix #10801

### Review suggestion

1. Please go through the discussion https://github.com/starwing/lua-protobuf/pull/240 before reviewing;
2. `*.pb.go` are auto-generated

### CODE REVIEW

File: spec/fixtures/grpc/targetservice.proto [hunk 1]
```diff
@@ -81,3 +88,8 @@ message Body {
   Limb legs = 3;
   Limb tail = 4;
 }
+
+message EchoMsg {
+  repeated string array = 1;
+  optional string nullable = 2;
+}
```
Comment (StarlightIbuki) [inline spec/fixtures/grpc/targetservice.proto:95]:
```suggestion
}

```

File: spec/helpers.lua [hunk 2]
```diff
@@ -1300,6 +1300,7 @@ end
 -- @return A thread object (from the `llthreads2` Lua package)
 -- @see kill_http_server
 local function http_server(port, opts)
+  print(debug.traceback("[warning] http_server is deprecated, use helpers.start_kong's fixture parameter or helpers.http_mock instead.", 2))
```
Comment (StarlightIbuki) [inline spec/helpers.lua:1303]:
```suggestion
```

**Review by chronolaw — APPROVED** [review]

**Comment by chronolaw** [comment]
Need a person familiar with Go to double check.

File: CHANGELOG.md [hunk 3]
```diff
@@ -190,6 +190,8 @@
   [#10663](https://github.com/Kong/kong/pull/10663)
 - **gRPC gateway**: `null` in the JSON payload caused an uncaught exception to be thrown during pb.encode.
   [#10687](https://github.com/Kong/kong/pull/10687)
+- **gRPC gateway**: fixed an issue that empty arrays in JSON response to be encoded as `"{}"`.
```
Comment (fffonion) [inline CHANGELOG.md:193]:
```suggestion
- **gRPC gateway**: fixed an issue that empty arrays in JSON are incorrectly encoded as `"{}"`; they are
now encoded as `"[]"` to comply with standard.
```

File: spec/03-plugins/28-grpc-gateway/01-proxy_spec.lua [hunk 4]
```diff
@@ -202,5 +202,35 @@ for _, strategy in helpers.each_strategy() do
       assert.equal(400, res.status)
     end)
 
+    describe("regression", function()
+      test("empty array in json #FTI-5002", function()
```
Comment (fffonion) [inline spec/03-plugins/28-grpc-gateway/01-proxy_spec.lua:206]: let's not use jira ticket number here, it's not visible to community and if we move to another jira project the number got missing. Let's open a github issue, describe the issue and use that instead.

**Review by fffonion — COMMENTED** [review]
two non-technical minor issues, otherwise lgtm

File: spec/03-plugins/28-grpc-gateway/01-proxy_spec.lua (hunk 4 above)
Comment (StarlightIbuki) [inline spec/03-plugins/28-grpc-gateway/01-proxy_spec.lua:206]:
```suggestion
      test("empty array in json #10801", function()
```

File: spec/03-plugins/28-grpc-gateway/01-proxy_spec.lua [hunk 5]
```diff
@@ -202,5 +202,35 @@ for _, strategy in helpers.each_strategy() do
...
+        assert.equal(200, res.status)
+  
+        local body = res:read_body()
+        assert.same(req_body, cjson.decode(body))
+        -- it should be encoded as empty array in json instead of `null` or `{}`
+        assert.matches("[]", body, nil, true)
+      end)
+  
+      -- Bug found when test FTI-5002's fix. It will be fixed in another PR.
+      pending("empty message #FTI-5054", function()
```
Comment (StarlightIbuki) [inline spec/03-plugins/28-grpc-gateway/01-proxy_spec.lua:221]:
```suggestion
      pending("empty message #10802", function()
```

**Review by flrgh — APPROVED** [review]

**Comment by flrgh** [comment]
rebased onto master to fix conflicts; good to merge whenever CI is :green_circle:

**Comment by kikito** [comment]
This fix was added very late so it was missing a cherrypick into 3.3 (notice that this PR targets master)

**Pull request merged**